### PR TITLE
Ruby: Add LDAP Injection query 

### PR DESCRIPTION
**Pull request by maikypedia**
This pull request adds a query for LDAP Injection covering [net-ldap](https://rubygems.org/gems/net-ldap).

Looking forward to your suggestions.

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/security/LdapInjectionCustomizations.qll [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  predicate isAdditionalTaintStep(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
+    filterTaintStep(nodeFrom, nodeTo)
+  }
+
+  private predicate filterTaintStep(DataFlow::Node n1, DataFlow::Node n2) {
+    exists(DataFlow::CallNode filterCall |
+      (
+        filterCall =
+          API::getTopLevelMember("Net").getMember("LDAP").getMember("Filter").getAMethodCall(["eq"]) or
+        filterCall.getMethodName() = ["[]"]
```
Comment (alexrford) [inline ruby/ql/lib/codeql/ruby/security/LdapInjectionCustomizations.qll:38]: I guess that this is for the `attributes` keyword arg in a case like: 
```rb 
ldap.search(base: "ou=people,dc=example,dc=com", filter: "cn=George", attributes: [name])
```
where we want to consider that a tainted element taints the whole array?

If so, we should maybe rename this predicate as `attributeArrayTaintStep` and make it such that it propagates flow from any indexed argument (`getArgument(_)`). It's also worth adding a line of documenting why we need this flow step.

File: ruby/ql/lib/codeql/ruby/security/LdapInjectionCustomizations.qll [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+   * A comparison with a constant string, considered as a sanitizer-guard.
+   */
+  private class StringConstCompareAsSanitizerGuard extends Sanitizer, StringConstCompareBarrier { }
+
+  /**
+   * An inclusion check against an array of constant strings, considered as a
+   * sanitizer-guard.
+   */
+  private class StringConstArrayInclusionCallAsSanitizer extends Sanitizer,
+    StringConstArrayInclusionCallBarrier { }
```
Comment (alexrford) [inline ruby/ql/lib/codeql/ruby/security/LdapInjectionCustomizations.qll:67]: Perhaps also add https://www.rubydoc.info/gems/net-ldap/Net/LDAP/Filter#escape-class_method as a sanitizer?

File: ruby/ql/lib/codeql/ruby/security/LdapInjectionQuery.qll [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+private import codeql.ruby.DataFlow
+private import codeql.ruby.TaintTracking
+private import LdapInjectionCustomizations
+private import LdapInjectionCustomizations::LdapInjection
+
+/**
+ * A taint-tracking configuration for detecting LDAP Injections vulnerabilities.
+ */
+class Configuration extends TaintTracking::Configuration {
```
Comment (alexrford) [inline ruby/ql/lib/codeql/ruby/security/LdapInjectionQuery.qll:14]: Would you mind converting this to a `DataFlow::ConfigSig` as in https://github.com/github/codeql/pull/13130/files#diff-cb35acf3aed06823bd9b005af3d73697a4cb92867d123eb7b868f9b6143c4082 (using `isBarrier` and `isAdditionalFlowStep` in place of `isSanitizer` and `isAdditionalTaintStep` respectively).

**Review by alexrford — CHANGES_REQUESTED** [review]
Thanks for the query! I'm overall happy with this, just some minor requests.

File: ruby/ql/lib/codeql/ruby/security/LdapInjectionQuery.qll [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+
+    predicate isBarrier(DataFlow::Node node) { node instanceof Sanitizer }
+
+    predicate isAdditionalFlowStep(DataFlow::Node node1, DataFlow::Node node2) {
+      LdapInjection::isAdditionalFlowStep(node1, node2)
+    }
+  }
+
+  import TaintTracking::Make<Config>
```
Comment (alexrford) [inline ruby/ql/lib/codeql/ruby/security/LdapInjectionQuery.qll:28]:
```suggestion
  import TaintTracking::Global<Config>
```

File: ruby/ql/src/experimental/ldap-injection/LdapInjection.ql [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+ * @tags security
+ *       external/cwe/cwe-090
+ */
+
+import codeql.ruby.DataFlow
+import codeql.ruby.security.LdapInjectionQuery
+import LdapInjection::PathGraph
+
+from LdapInjection::PathNode source, LdapInjection::PathNode sink
+where LdapInjection::hasFlowPath(source, sink)
```
Comment (alexrford) [inline ruby/ql/src/experimental/ldap-injection/LdapInjection.ql:19]:
```suggestion
where LdapInjection::flowPath(source, sink)
```

**Review by alexrford — CHANGES_REQUESTED** [review]
These changes should fix the deprecation warnings.

**Comment by alexrford** [comment]
Hey @maikypedia, sorry for the delay on this PR. I had another look through this PR and there are a few comments from my initial review that got lost in the changes. Would you mind looking through the comments in that review again?

**Comment by maikypedia** [comment]
> Hey @maikypedia, sorry for the delay on this PR. I had another look through this PR and there are a few comments from my initial review that got lost in the changes. Would you mind looking through the comments in that review again?

Sorry, I completely overlooked it 😅. I think with this, everything should be ready now. Is there anything I might be missing?

**Review by alexrford — APPROVED** [review]

**Pull request merged**